### PR TITLE
Adds support for Mapzen terrain

### PR DESCRIPTION
**Pull request by ibesora**
## Launch Checklist
 - [x] briefly describe the changes in this PR
Implements #5824 and adds support for the Mapzen terrarium terrain format by adding a new encoding parameter to the raster-dem source type that selects which unpack function should be used.
Also adds a new debug map that shows this new terrain format.
 - [ ] write tests for all new functionality
 - [x] document any changes to public APIs
Adds a new parameter called _encoding_ to the raster-dem source type. Defaults to mapbox to use the mapbox terrain encoding
 - [ ] post benchmark scores
 - [x] manually test the debug page
Works as expected

### CODE REVIEW

File: src/style-spec/reference/v8.json [hunk 1]
```diff
@@ -231,7 +231,7 @@
       "type": "enum",
       "values": {
         "raster-dem": {
-            "doc": "A raster DEM source using Mapbox Terrain RGB"
+            "doc": "A raster DEM source"
```
[the remark this answers is not in the document]
Comment (ibesora) [reply src/style-spec/reference/v8.json]: Corrected it in the new commit. 
I also expanded the declaration of the encoding parameter to explicitly tell the supported values

File: src/data/dem_data.js [hunk 2]
```diff
@@ -57,22 +57,14 @@ class DEMData {
...
-        // unpack
-        for (let y = 0; y < level.dim; y++) {
-            for (let x = 0; x < level.dim; x++) {
-                const i = y * level.dim + x;
-                const j = i * 4;
-                // decoding per https://blog.mapbox.com/global-elevation-data-6689f1d0ba65
-                level.set(x, y, this.scale * ((pixels[j] * 256 * 256 + pixels[j + 1] * 256.0 + pixels[j + 2]) / 10.0 - 10000.0));
-            }
-        }
+        this.unpackData(level, data, encoding);
```
Comment (mollymerp) [inline src/data/dem_data.js]: needs to be `data.data`

File: src/data/dem_data.js [hunk 3]
```diff
@@ -57,22 +57,14 @@ class DEMData {
         this.loaded = !!data;
     }
 
-    loadFromImage(data: RGBAImage) {
+    loadFromImage(data: RGBAImage, encoding: string) {
         if (data.height !== data.width) throw new RangeError('DEM tiles must be square');
 
         // Build level 0
         const level = this.level = new Level(data.width, data.width / 2);
         const pixels = data.data;
```
Comment (mollymerp) [inline src/data/dem_data.js]: can remove `pixels` or use it in L67

File: src/style-spec/reference/v8.json [hunk 4]
```diff
@@ -272,6 +272,10 @@
       "type": "string",
       "doc": "Contains an attribution to be displayed when the map is shown to a user."
     },
+    "encoding": {
+      "type": "string",
```
Comment (mollymerp) [inline src/style-spec/reference/v8.json]: this should be `"type": "enum"` with a `values` object listing the possible types.

File: src/source/raster_dem_tile_source.js [hunk 5]
```diff
@@ -15,11 +15,14 @@ import type {Callback} from '../types/callback';
...
 
 class RasterDEMTileSource extends RasterTileSource implements Source {
+    encoding: string;
+
     constructor(id: string, options: RasterDEMSourceSpecification, dispatcher: Dispatcher, eventedParent: Evented) {
         super(id, options, dispatcher, eventedParent);
         this.type = 'raster-dem';
         this.maxzoom = 22;
         this._options = util.extend({}, options);
+        this.encoding = "mapbox";
```
Comment (mollymerp) [inline src/source/raster_dem_tile_source.js]: `options.encoding || "mapbox";`

File: src/data/dem_data.js [hunk 6]
```diff
@@ -95,6 +87,40 @@ class DEMData {
...
+            }
+        }
+    }
+
+    unpackMapzenData(level: Level, pixels: Uint8Array) {
+        for (let y = 0; y < level.dim; y++) {
+            for (let x = 0; x < level.dim; x++) {
+                const i = y * level.dim + x;
+                const j = i * 4;
+                // decoding per https://mapzen.com/documentation/terrain-tiles/formats/#types-of-terrain-tiles
```
Comment (mollymerp) [inline src/data/dem_data.js]: link to AWS docs

**Comment by ibesora** [comment]
Hi @mollymerp . 
When looking at your fixes I saw that I had not pushed my last commit that corrected some of the things you told me :'(
I've pushed my changes, added the AWS endpoint, changed the encoding type to an enum and added a toggle to the debug page

File: src/source/raster_dem_tile_source.js [hunk 7]
```diff
@@ -15,11 +15,14 @@ import type {Callback} from '../types/callback';
 
 
 class RasterDEMTileSource extends RasterTileSource implements Source {
+    encoding: string;
```
Comment (mollymerp) [inline src/source/raster_dem_tile_source.js]: you can explicitly type this as `encoding: "mapbox" | "terrarium";`

File: src/data/dem_data.js [hunk 8]
```diff
@@ -95,6 +87,40 @@ class DEMData {
         this.loaded = true;
     }
 
+    unpackData(level: Level, pixels: Uint8Array | Uint8ClampedArray, encoding: string) {
```
Comment (mollymerp) [inline src/data/dem_data.js]: to reduce code duplication, you could refactor to something like
```js
unpackMapbox(x, y, r, g, b) {
  level.set(x, y, this.scale * ((r * 256 * 256 +g * 256.0 + b) / 10.0 - 10000.0));
}
unpackTerrarium(x, y, r, g, b) {
  level.set(x, y, this.scale * ((r * 256 + g+ b / 256) - 32768.0));
}
const unpackFunctions = {"mapbox": unpackMapbox, "terrarium": unpackTerrarium};
const unpack = unpackFunctions[encoding];
 for (let y = 0; y < level.dim; y++) {
   for (let x = 0; x < level.dim; x++) {
      const i = y * level.dim + 
      const j = i * 4;
      unpack(x, y, pixels[j], pixels[j+1], pixels[j+2]);
  }
}
```

File: src/source/worker_source.js [hunk 9]
```diff
@@ -28,7 +28,8 @@ export type WorkerTileParameters = TileParameters & {
 
 export type WorkerDEMTileParameters = TileParameters & {
     coord: { z: number, x: number, y: number, w: number },
-    rawImageData: RGBAImage
+    rawImageData: RGBAImage,
+    encoding: string
```
Comment (mollymerp) [inline src/source/worker_source.js]: explicitly type for enum values (as above)

File: src/style-spec/reference/v8.json [hunk 10]
```diff
@@ -272,6 +272,19 @@
...
       "doc": "Contains an attribution to be displayed when the map is shown to a user."
     },
+    "encoding": {
+      "type": "enum",
+      "values": {
+        "terrarium": {
+          "doc": "Terrarium format PNG tiles. See https://aws.amazon.com/es/public-datasets/terrain/ for more info."
+        },
+        "mapbox": {
+          "doc": "Mapbox Terrain RGB tiles. See https://www.mapbox.com/blog/terrain-rgb/ for more info."
```
Comment (mollymerp) [inline src/style-spec/reference/v8.json]: I just found out we have official documentation for this now, so we can link to https://www.mapbox.com/help/access-elevation-data/#mapbox-terrain-rgb

**Comment by ibesora** [comment]
Hi again @mollymerp I've just pushed some changes that incorporate your comments.

File: src/data/dem_data.js [hunk 11]
```diff
@@ -57,22 +57,14 @@ class DEMData {
         this.loaded = !!data;
     }
 
-    loadFromImage(data: RGBAImage) {
+    loadFromImage(data: RGBAImage, encoding: ?string) {
```
Comment (mollymerp) [inline src/data/dem_data.js]: this typing can also be `"mapbox | "terrarium"` – sorry I missed it earlier!

File: src/data/dem_data.js [hunk 12]
```diff
@@ -95,6 +87,26 @@ class DEMData {
         this.loaded = true;
     }
 
+    unpackMapbox(r: number, g: number, b: number) {
```
Comment (mollymerp) [inline src/data/dem_data.js]: I think we can use `_unpack*` for these functions to signify that they're private to the `DEMData` class

File: src/data/dem_data.js (hunk 11 above)
Comment (ibesora) [reply src/data/dem_data.js]: Done

File: src/data/dem_data.js (hunk 12 above)
Comment (ibesora) [reply src/data/dem_data.js]: Done!

**Comment by mollymerp** [comment]
@ibesora thanks for your patience and great work here! to save you the trouble I have rebased and fixed conflicts and added some tests. I had to push to a new branch so I've made a PR for that, but I think this is good to go overall! 👉  #6110

**Comment by mollymerp** [comment]
closing here in favor of #6110

**Pull request closed without merging**